lib/spack/spack/object_file/elf.py

```diff
@@ -0,0 +1,293 @@
+        self.dt_offset = dt_offset
+        self.is_runpath = is_runpath
+
+
+def get_rpaths_byte_string_and_offset(f):
+    """
+    Given a file handle f for an ELF file, return the rpath as a (byte) string, as well
+    as the offset into the file at which it occurs, in case it has to be overwritten
+    in-place later.
+    """
+    # Go to the e_ident header
+    f.seek(0)
+
+    # Read the 32/64 bit class independent part of the header and validate
+    e_ident = f.read(16)
+
+    # Require ELF magic bytes.
+    if len(e_ident) != 16 or e_ident[:4] != b'\x7fELF':
+        raise ElfParsingError("No ELF file")
+
+    # Defensively require a valid class and data.
+    e_ident_class, e_ident_data = get_byte_at(e_ident, 4), get_byte_at(e_ident, 5)
+    if e_ident_class != 0x1 and e_ident_class != 0x2:
+        raise ElfParsingError("Invalid class found")
+    if e_ident_data != 0x1 and e_ident_data != 0x2:
+        raise ElfParsingError("Invalid data type")
+
+    is_64_bit = e_ident_class == 0x02
+    is_little_endian = e_ident_data == 0x01
+
+    # Set up byte order and types for unpacking
+    bo = '<' if is_little_endian else '>'
+    t_half_half_half_half = bo + 'HHHH'
+    t_word = bo + 'I'
+    if is_64_bit:
+        t_off = bo + 'Q'
+        t_off_addr = bo + 'QQ'
+        t_off_off = t_off_addr
+        t_Dyn = bo + 'qQ'
+
+        header_remainder_size = 48
+        e_phoff_offset = 16
+        e_phentsize_offset = 38
+        p_offset_offset = 8
+        Dyn_size = 16
+        sh_offset_offset = 24
+
+    else:
+        t_off = bo + 'L'
+        t_off_addr = bo + 'LL'
+        t_off_off = t_off_addr
+        t_Dyn = bo + 'lL'
+
+        e_phoff_offset = 12
+        header_remainder_size = 36
+        e_phentsize_offset = 26
+        p_offset_offset = 4
+        Dyn_size = 8
+        sh_offset_offset = 16
+
+    # read the remaining bytes of the header
+    header = f.read(header_remainder_size)
+    if len(header) != header_remainder_size:
+        raise ElfParsingError("ELF header malformed")
+
+    e_type = get_byte_at(header, 0)
+
+    if e_type != 0x02 and e_type != 0x03:
+        raise ElfParsingError("Not an ET_DYN or ET_EXEC type")
+
+    # Find program header and segment header offsets, size and count.
+    e_phoff, e_shoff = unpack_from(
+        t_off_off, buffer=header, offset=e_phoff_offset)
+    e_phentsize, e_phnum, e_shentsize, e_shnum, = unpack_from(
+        t_half_half_half_half, buffer=header, offset=e_phentsize_offset)
+
+    # go over the sections
+    f.seek(e_phoff)
+
+    # Here we have to make a mapping from virtual address to offset in the file.
+    offsets = []
+    virtual_addresses = []
+
+    found = False
+    for _ in range(e_phnum):
+        prog_header = f.read(e_phentsize)
+        if len(prog_header) != e_phentsize:
+            raise ElfParsingError("Malformed program header")
+        p_type, = unpack_from(t_word, buffer=prog_header, offset=0)
+
+        # Find PT_LOAD to get the
```

Comment:
```suggestion
        # For PT_LOAD entries: Save the offsets and virtual addrs of the loaded ELF segments
```
(Unless my guess is wrong our you know a better way to comment this)